### PR TITLE
run psalm with all NC versions

### PR DESCRIPTION
**Pull request by individual-it**
## Description

This PR following activities are performed: 
- fixes the error while running psalm 

```console
PHP Fatal error:  Declaration of Doctrine\Common\Cache\Psr6\CacheItem::get() must be compatible with Psr\Cache\CacheItemInterface::get(): mixed in nextcloud/master/3rdparty/doctrine/cache/lib/Doctrine/Common/Cache/Psr6/CacheItem.php on line 51
```

- Removing the phpstan from the CI

## Related Issue or Workpackage

- https://community.openproject.org/projects/nextcloud-integration/work_packages/50516/activity

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Updated `CHANGELOG.md` file

### CODE REVIEW

File: .github/workflows/shared_workflow.yml [hunk 1]
```diff
@@ -72,13 +72,16 @@ jobs:
           cd server && git submodule update --init
           ./occ maintenance:install --admin-pass=admin
 
-      - name: PHP code analysis
+      - name: PHP code analysis for phpstan
```
Comment (saw-jan) [inline .github/workflows/shared_workflow.yml:75]:
```suggestion
      - name: PHP code analysis - phpstan
```

File: .github/workflows/shared_workflow.yml [hunk 2]
```diff
@@ -72,13 +72,16 @@ jobs:
...
         run: |
           if [ ${{ matrix.nextcloudVersion }} == "stable26" -o ${{ matrix.nextcloudVersion }} == "stable27" ]
           then
             make phpstan
           fi
 
+      - name: PHP code analysis for psalm
```
Comment (saw-jan) [inline .github/workflows/shared_workflow.yml:82]:
```suggestion
      - name: PHP code analysis - psalm
```

**Comment by SagarGi** [comment]
@nabim777 If we are able to make run psalm with all version of NC thenlets remove `phpstan` since we can achieve the same thing with psalm. I would suggest to remove `phpstan` from CI in this PR and make a workpackage to clear the things up with `phpstan`.
CC @individual-it

File: .github/workflows/shared_workflow.yml [hunk 3]
```diff
@@ -76,7 +76,9 @@ jobs:
         run: |
           if [ ${{ matrix.nextcloudVersion }} == "stable26" -o ${{ matrix.nextcloudVersion }} == "stable27" ]
           then
-            make phpstan
+            make psalm
+          else
+            make psalm
           fi
```
Comment (SagarGi) [inline .github/workflows/shared_workflow.yml:82]: just `make psalm`

File: tests/stub.phpstub [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+		public function expiresAfter($time): self
+		{
+		}
+
+		public function getExpiry(): ?float
+		{
+		}
+	}
+}
```
Comment (SagarGi) [inline tests/stub.phpstub:36]: can you give a comment what's this actually needed for? can be useful in the future when someone come across it.

**Review by SagarGi — CHANGES_REQUESTED** [review]

File: tests/stub.phpstub [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
```
Comment (nabim777) [inline tests/stub.phpstub:22]:
```suggestion
declare(strict_types=1);
/**
 * @copyright Copyright (c) 2024 Nabin Ale <nabinalemagar019@gmail.com>
 *
 * @license GNU AGPL version 3 or any later version
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
```
Other apps stub file contains this so should I need to add this or not here?
https://github.com/nextcloud/groupfolders/blob/master/tests/stub.phpstub#L5
https://github.com/nextcloud/user_retention/blob/main/tests/stub.phpstub#L4

**Review by SagarGi — APPROVED** [review]
LGTM :+1:

File: tests/stub.phpstub (hunk 5 above)
Comment (SagarGi) [reply tests/stub.phpstub]: Can be added.

File: psalm.xml [hunk 6]
```diff
@@ -9,24 +9,35 @@
...
+	</projectFiles>
+	<stubs>
+		<file name="tests/stub.phpstub" preloadClasses="true"/>
+	</stubs>
 	<issueHandlers>
 		<UndefinedClass>
 			<errorLevel type="suppress">
 				<referencedClass name="OC\Http\Client\LocalAddressChecker" />
-				<!-- these are classes form activity app, which isn't cloned while doing static code analysis -->
+				<!-- this is a class form terms_of_service app, which isn't cloned while doing static code analysis -->
```
Comment (individual-it) [inline psalm.xml:26]: It actually might be cloned, so I would rephrase the comment to something like:
```suggestion
				<!-- these classes belong to the terms_of_service app, which isn't compulsory, so might not exist while running psalm -->
```

same for other comments about missing classes

File: lib/AppInfo/Application.php [hunk 7]
```diff
@@ -106,10 +106,13 @@ public function boot(IBootContext $context): void {
 		$dispatcher->addServiceListener(BeforeUserDeletedEvent::class, BeforeUserDeletedListener::class);
 		$dispatcher->addServiceListener(BeforeGroupDeletedEvent::class, BeforeGroupDeletedListener::class);
 		$dispatcher->addServiceListener(UserChangedEvent::class, UserChangedListener::class);
+		/** @psalm-suppress InvalidArgument */
```
Comment (individual-it) [inline lib/AppInfo/Application.php:109]: can we use normal single line comments?
`// @psalm-suppress InvalidArgument `
and also please add a comment about why we are suppressing this error

File: .github/workflows/shared_workflow.yml [hunk 8]
```diff
@@ -73,11 +73,7 @@ jobs:
           ./occ maintenance:install --admin-pass=admin
 
       - name: PHP code analysis
-        run: |
-          if [ ${{ matrix.nextcloudVersion }} == "stable26" -o ${{ matrix.nextcloudVersion }} == "stable27" ]
-          then
-            make phpstan
```
Comment (individual-it) [inline .github/workflows/shared_workflow.yml:79]: the PR description does not say anything about removing phpstan. Should that happen in this PR or is a separate see https://community.openproject.org/work_packages/55383/activity ?

Comment (nabim777) [reply .github/workflows/shared_workflow.yml:79]: The reason behind removing the PHPstan from CI  is that there is a conflict on code analyzer between Psalm and PHPstan on CI and locally. https://github.com/nextcloud/integration_openproject/actions/runs/9284704953/job/25547647168 
In another PR, PHPstan will definitely removed so there will be just waste of time solving the conflict  between the two code analyzer in this PR. So, in this PR there is remove of the PHPstan from the CI and completely remove in another PR. Will be this Okay @individual-it ?

File: lib/AppInfo/Application.php (hunk 7 above)
Comment (nabim777) [reply lib/AppInfo/Application.php]: using  `// @psalm-suppress InvalidArgument ` does not work. Could we use `/** @psalm-suppress InvalidArgument */` like other apps? @individual-it 
https://github.com/nextcloud/groupfolders/blob/b6fdec2266af3f40d50c0c90f42b5a53b97c7dbb/lib/Command/Scan.php#L53

https://github.com/nextcloud/all-in-one/blob/d6580b30ceccabad473fc519966e20b385d7c2dc/php/src/Docker/DockerActionManager.php#L756

**Review by individual-it — COMMENTED** [review]
LGTM
